### ISSUE TITLE
Broad question about Hypothesis

### ISSUE THREAD

**Issue by rsokl** [HypothesisWorks/hypothesis#1641]
I have been using hypothesis for some educational tools that I am developing, and would like to give a presentation to my colleagues about the Hypothesis libraries.

While doing a literature review about various testing methodologies (QuickCheck, AFL, the swarm testing that is discussed in #1637, etc) to learn more about this area of work, I realized that I have no idea if the engine driving Hypothesis stems from any of these works. Are there any papers that inspired Hypothesis' present approach to its search strategies? 

I guess I am really trying to understand where Hypothesis sits relation to other works in this field. Perhaps the answer is as succinct as me needing to learn about monadic parser combinators :smile: 

Also, thank you, @DRMacIver and @Zac-HD,  very much for all of your incredible work. It would have been impossible for me to implement a trustworthy [autograd library](https://github.com/rsokl/MyGrad) for my students to learn from and contribute to if it weren't for Hypothesis. 

([Here is a cool rule-based state machine that I created](https://github.com/rsokl/MyGrad/blob/master/tests/state_testing/test_state.py) to ensure that back-propagation through my topologically-sorted computational graph matches a brute force implementation)

**Comment by DRMacIver** [comment]
The main research Hypothesis is built on is QuickCheck. Roughly Hypothesis is QuickCheck + One Big Idea (which is the monadic parser combinators stuff. I hope to be able to announce a paper about that soon, though we never use the term in this paper. Will know in about a month. 🤞) + A lot of UX and engineering work.

It's had a lot of "general influence" from studying the design of fuzzers like American Fuzzy Lop, and the work of the compiler testing community on things like Csmith and C-Reduce, but it would be hard to point to any specific features of Hypothesis that come from those so much as that they're good places to learn about this sort of testing.

**Comment by DRMacIver** [comment]
Oh, and PS, thanks for the kind words! Autograd looks like a neat library, and that's a cool application of the stateful testing. Did it find anything fun?

**Comment by rsokl** [comment]
Thank you for the feedback (and for taking care to use the full names of the various fuzzers so that I can read up on them)! I'll be eager to read that paper if/when it comes through.

The stateful testing absolutely found some fun stuff. There were probably 10 problematic combinations of graph topologies and backprop/clear-gradient calls that it sniffed out. My code actually got simpler and simpler as I used the stateful testing to drive the code's development and to fix my bad logic - it was a ton of fun.

My unit tests with Hypothesis were so effective that [I found a bug in numpy](https://github.com/numpy/numpy/issues/10930) and several in Harvard's ([.kinda.](https://github.com/HIPS/autograd/issues/376) [.crappy.](https://github.com/HIPS/autograd/issues/378)) autograd library.

**Comment by DRMacIver** [comment]
Heh. I wish you'd got in touch yesterday. I would have included some of this in a talk I gave to a bunch of neural network specialists this afternoon. This is really neat either way though, thanks for sharing!

**Comment by Zac-HD** [comment]
Thanks for getting in touch Ryan!  It's always lovely to hear from happy users doing cool things with Hypothesis, and that's a very nice trophy case 😄

Closing the issue because there's no specific action we need to take, but I'd love to stay in touch!
